Issue #5306: add additional message for void return

vaano94 opened the pull request:
Issue #5306

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -199,9 +204,11 @@ private void visitReturn(DetailAST ast) {
         // after the first return statement is seen
         if (ast.getFirstChild().getType() == TokenTypes.SEMI) {
+            context.hasVoidReturn = true;
             context.visitLiteralReturn(maxForVoid);
```
usually accessing a field directly is frowned upon.
We have the `visit` method call right below the `hasVoidReturn`, so we should pass the value to that method instead.

vaano94 replied:
done

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -199,9 +204,11 @@ private void visitReturn(DetailAST ast) {
         }
         else {
+            context.hasVoidReturn = false;
             context.visitLiteralReturn(max);
```
same here.

vaano94 replied:
done

rnveach commented:
@vaano94 Please read over https://github.com/checkstyle/checkstyle/wiki/PR-rules especially point 2 and 6.
There should only be 1 commit in this PR.

vaano94 commented:
@rnveach Thank you very much. I squashed commits in one. 
@Vladlis I renamed the field into 'isVoidContext' as we have discussed and provided additional doc.

Vladlis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -217,6 +222,8 @@ private void visitReturn(DetailAST ast) {
         private int count;
         /** Maximum allowed number of return statements. */
         private Integer maxAllowed;
+        /** Identifies if context (method/lambda/constructor) returns anything. */
```
This javadoc makes me feel like if `isVoidContext` is `true` then context returns something, but it should have an oppposite meaning. Please a change to `Identifies if context is void.` - I think that is enough clear enough.

vaano94 replied:
corrected

Vladlis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -227,10 +234,12 @@ private void visitReturn(DetailAST ast) {
 
         /**
-         * Increase the number of return statements.
+         * Increase the number of return statements and set method return type.
```
Please use `context` instead of `method`

vaano94 replied:
done

Vladlis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -227,10 +234,12 @@ private void visitReturn(DetailAST ast) {
-         * Increase the number of return statements.
+         * Increase the number of return statements and set method return type.
          * @param maxAssigned Maximum allowed number of return statements.
+         * @param voidReturn Identifies if context (method/lambda/constructor) returns anything.
```
Please use `Identifies if context is void.`

vaano94 replied:
done

Vladlis commented on src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages.properties:
```diff
@@ -44,7 +44,8 @@ overload.methods.declaration=Overload methods should not be split. Previous over
 require.this.variable=Reference to instance variable ''{0}'' needs \"{1}this.\".
-return.count=Return count is {0,number,integer} (max allowed is {1,number,integer}).
+return.count=Return count is {0,number,integer} (max allowed for non-void methods/lambdas is {1,number,integer}).
+return.countVoid=Return count is {0,number,integer} (max allowed for void methods/constructors/lambdas methods is {1,number,integer}).
```
`for void methods/constructors/lambdas methods is` - word `methods` is repeated twice, please fix here and in all translations.

vaano94 replied:
done. In all further translations it is not repeated.

vaano94 commented:
@Vladlis Could you please have a check on my last commit?

Vladlis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -227,10 +234,12 @@ private void visitReturn(DetailAST ast) {
          */
-        public void visitLiteralReturn(int maxAssigned) {
+        public void visitLiteralReturn(int maxAssigned, Boolean voidReturn) {
+            context.isVoidContext = voidReturn;
```
Why do you set property to the outer class field? Please refer directly to `isVoidContext` from inside the same object:
```java
isVoidContext = voidReturn;
```

vaano94 replied:
done

Vladlis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -245,7 +254,12 @@ public void visitLiteralReturn(int maxAssigned) {
         public void checkCount(DetailAST ast) {
             if (checking && maxAllowed != null && count > maxAllowed) {
-                log(ast.getLineNo(), ast.getColumnNo(), MSG_KEY, count, maxAllowed);
+                if (context.isVoidContext) {
```
Same here, please use `isVoidContext` instead of `context.isVoidContext`

vaano94 replied:
done

Vladlis commented on src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages_pt.properties:
```diff
@@ -44,7 +44,8 @@ overload.methods.declaration=Métodos de sobrecarga não devem ser separados. O
 require.this.variable=A referência à variável de instância ''{0}'' precisa de \"{1}this.\".
-return.count=O número de \"return\"s é {0,number,integer} (o máximo permitido é {1,number,integer}).
+return.count=O número de \"return\"s é {0,number,integer} (o máximo permitido para métodos não vazios/lambdas é {1,number,integer}).
+return.countVoid=O número de \"return\"s é {0,number,integer} (o máximo permitido para métodos vazios/construtores/métodos/lambdas é {1,number,integer}).
```
`métodos` is used twice, please recheck all translations for duplicating `methods`

vaano94 replied:
done
